ARCs/arc-0039.md

```diff
@@ -0,0 +1,54 @@
+---
+arc: 39
+title: Dynamic Tuple End In Head
+description: Append the end of a dynamic tuple to the head values
+author: Joe Polny (@joe-p)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/212
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-07-15
+requires: 4
+---
+
+## Abstract
+According to [ARC-0004](./arc-0004.md), when encoding dynamic types in a tuple the tuple consists of a `head` containing all of the offsets of the actual values in the `tail`. An additional value, the byte offset of the end of the tuple, should be added to the end of the `head` value for more efficient element reading.
+
+## Motivation
+When reading a value in a dynamic tuple, the head offset of the element must be extracted as a `uint16` to get the start of the element. To get the end of the element, which is necessary to properly read the element, the following head value can be extracted. This, however, fails to work when the element being read is the last element in its respective dynamic array or tuple. When the element is the last in the array or tuple, one must compute the length of the element. For static types, this is trivial and for dynamic arrays of static types this is also trivial because the length prefix can be used. For more complex nested dynamic types, however, this could involve many levels of extracting lengths and offsets, thus reading an element is `O(N)`, which `N` being the depth of the nested dynamic types. If there was an additional head value containing the end of the dynamic tuple, readnig any element, regardless of its type or position, would be `O(1)` since one can always just extract the following head value.
```

Comment:
> which is necessary to properly read the element

I don't see why this is true. Grab the suffix of the input, beginning at the start of the value.  The value itself contains everything you need to read the right amount, starting there.  There's no particular reason to worry, ahead of time, where the value ends, so just pass the entire suffix to whatever needs to decode it.

Reply:
If the value is dynamic it doesn't contain "everything you need". If the value is a dynamic array the length prefix will tell you how many elements are in the array but not the length of those elements. You need to read the last value of the array and get the length for that. And if that value is a dynamic array, you must do the same thing.

Reply:
Even if the elements are dynamic, their heads are statically sized, and laid out first.  You know how many there are from the length.  And, if they are dynamic, they will have pointers to their dynamically sized tails.

Reply:
But in the case of nested dynamic types you have to continuously extract the head and calculate length. If we knew the end of the tuple it'd be much easier to use that to extract/substring. 

To be clear, the desired goal here is to be able to efficiently extract a single value from an encoded nested dynamic type.